Refactor DPU Tools into a uniform interface

SamD2021 opened the pull request:
- All scripts have been either turned into modules, or been housed in existing modules.
- Merged Dockerfiles together.
- Added a new DPU Types enum class.
- Added DPU Hardware Detection.
- Refactored DPU Tools script into a class, since there were many variables that should be shared between functions
- Transformed pxeboot script into a module and  refactored it into a class.
- Added firmware class for Bluefields into fwutils module.
- Made capture output the default for the run function in the common module.
- Most tools can be utilized with auto hardware detection
- Added Console check for DPU Hardware Type

SamD2021 committed:
Reorganize the tools for a unified interface

This is in preparation for making a unified interface for dpu tools that
is able to detected the hardware its running on an call its respective
tools

ipu: Remove IPU dir

SamD2021 committed:
dpu-tools: Modify ipu and common related imports

Make the imports match the new structure

SamD2021 committed:
Move common hardware agnostic logic from common_ipu.py into common.py

Many of these functions aren't necessarily tied to the ipu so moving it
to the common module makes more sense

SamD2021 committed:
dpu-tools: import setup_logging and run from common.py

SamD2021 committed:
Dockerfile: Merge the Dockerfiles into one

Since we want to unify the scripts behind dpu-tools we should also build
the tools using the same Dockerfile

SamD2021 committed:
dpu-tools: Remove redundant code

SamD2021 committed:
ipu/fwutils.py: Update import to match new structure

Matching imports to new structure

SamD2021 committed:
Move console function from dpu-tools script to common_ipu.py

SamD2021 committed:
Move console function from bf/console script to common_bf.py

SamD2021 committed:
dpu-tools: Use Hardware detection for console

SamD2021 committed:
dpu-tools: Refactor DPUTools into a class

Since DPU tools often relies passing the args to each function, it would
be nice to have a data structure to bundle all of this in. Likewise it
would be good if we could also keep track of what DPU type are working
with.

SamD2021 committed:
common_ipu.py: Change console method to console_ipu

SamD2021 committed:
utils/fwutils.py: Adapt imports to match package structure

SamD2021 committed:
entry.sh: Add a script to start rshim and run dpu-tools

This script is useful because it lets rshim run in the background, which
the Bluefields DPUs need to console into them.

SamD2021 committed:
Dockerfile: Consolidate rshim logic into an entry script

SamD2021 committed:
utils/common.py: Add DPU Types

This enum keeps us with a structured way to track different DPU types.
Previously we would just use hardcoded strings, but with this class we
can even validate if the DPU Type is within our tracked types

SamD2021 committed:
utils/common.py: Add scan_for_dpus

SamD2021 committed:
utils/common.py: Add DPU Hardware detection

SamD2021 committed:
Move list_dpus into dpu-tools

SamD2021 committed:
utils/common_ipu.py: Fix imports

SamD2021 committed:
utils/common_ipu.py: Use configure minicom for creating ipu console as well

SamD2021 committed:
utils/fwutils.py: Add BFFirmware

By making this class we can keep together related logic to reset,
upgrade and version firmware across dpu types

SamD2021 committed:
utils/fwutils.py: Add firmware version for BF

SamD2021 committed:
utils/remote_api.py: Move remote api class into its own module

SamD2021 committed:
utils/fwutils.py: Add Firmware Upgrade for BF

SamD2021 committed:
utils/fwutils.py: Add Firmware Reset for BF

SamD2021 committed:
dpu-tools: Add Hardware Detection for Firmware Version

SamD2021 committed:
dpu-tools: Add Hardware detection for required flags

SamD2021 committed:
dpu-tools: Consume Firmware Reset for BF

SamD2021 committed:
dpu-tools: Add BF option for Firmware Up

SamD2021 committed:
utils/common.py: Make capture_output default to true

Many of the Bluefield scripts requires capturing the output. It would be
better to have it default to capturing output and have an option to turn
it off when we think its not useful

SamD2021 committed:
utils/common_bf.py: Use the run function from the common module

SamD2021 committed:
utils/common_bf.py: Add get_mode

SamD2021 committed:
dpu-tools: Consume get_mode in the main interface

SamD2021 committed:
utils/common_bf.py: Add set_mode

SamD2021 committed:
dpu-tools: Consume set_mode in the main interface

Here we are using a single mode subcommand that can be used to retrieve
the mode and set if passed the set-mode flag

SamD2021 committed:
utils/common_bf.py: Move functionality to download BFB images to the BF

SamD2021 committed:
utils/common_bf.py: Move functionality to reset BF into its common module

SamD2021 committed:
Move and convert pxeboot into a utils module

SamD2021 committed:
dpu-tools: Add command to interact with PXEBOOT logic

SamD2021 committed:
dpu-tools: Let the reset command call the reset function which can call BF reset

SamD2021 committed:
dpu-tools: Properly validate the console subcommand

SamD2021 committed:
Remove redundant bf package

Remove redundant fwversion

Remove redundant fwup

Remove redundant fwdefaults

Remove listbf script

The list bf script has been integrated into the list dpus command and
function. Now we can use this information to auto detect hardware and
have the tool behave accordingly depending on its hardware

SamD2021 committed:
utils/fwutils.py: Add Firmware Upgrade for cx

SamD2021 committed:
dpu-tools: Consume Firmware Upgrade for CX Cards

SamD2021 committed:
dpu-tools: Fix main function return annotation

SamD2021 committed:
mypy.ini: Adjust mypy to match new structure and imports

SamD2021 committed:
dpu-tools: Consume bfb for BF

SamD2021 committed:
README.md: Update Readme with new guide to interact with the tools

SamD2021 committed:
pyproject.toml: Match new structure

SalDaniele commented on README.md:
```diff
@@ -1,31 +1,24 @@
 ## Tools
 
-All the tools can directly be ran inside the container. All the tools automatically find and act on the first BF in the system.
+All the tools can directly be interacted with through the dpu-tools interface. All the tools automatically find and act on the first DPU in the system. One small difference is that using the console subcommand requires specifying what DPU type you are working with. Check out its help page by passing the `-h` flag.
```
"one small difference" awkward wording, maybe something like "Note* specific subcommands require knowledge of the DPU type in question. For example, DPU-type is required when running the console command..."

SalDaniele commented on README.md:
```diff
@@ -1,31 +1,24 @@
-| `get_mode`   | Gets the BF mode.                                                                        |
-| `cx_fwup`    | Upgrades the firmware on a CX                                                            |
+| `mode`       | Gets the BF mode. Use `--set-mode` to change the mode to either dpu or nic               | 
+| `utils`      | Access common or non-dpu specific utilities. {cw_fwup, bfb}                              |
```
I like it, we just need to make sure to make an update accordingly in https://github.com/bn222/cluster-deployment-automation/blob/main/host.py#L428.

This applies to all the commands called from CDA.

SalDaniele commented on dpu-tools:
```diff
@@ -0,0 +1,278 @@
+
+    def reset(self) -> None:
+        if self.dpu_type == DPUType.IPU.name:
+            run("ssh root@100.0.0.100 sudo reboot")
```
We should update this to use the IMC address, this should be a required argument for any IPU commands that need connection to the IMC

SalDaniele commented on dpu-tools:
```diff
@@ -0,0 +1,278 @@
+logger = logging.getLogger(__name__)
+
+
+class DPUTools:
```
It might make sense to make this an abstract-base-class that each DPU derives from. Imagine you are adding a new DPU. You don't want to have to reach through every single line and add a bunch of new "if DPUType.my_new_dpu"... statements.

Might be nicer to do this in one place, and create a class object that inherits from DPUTools for each of these.

That said, this PR works and has been tested, so fine with adding this in a future PR.

SalDaniele commented on dpu-tools:
```diff
@@ -0,0 +1,278 @@
+
+
+def main() -> None:
+    detected = detect_dpu_type()
```
We should take the DPU type from the user if they provide it, otherwise we use the auto-detected dpu type.

We could even through a warning if the users dpu type and the autodetected do not match.

SalDaniele commented on dpu-tools:
```diff
@@ -0,0 +1,278 @@
+    )
+    bfb_parser.set_defaults(subcommand="bfb")
+    # Parse arguments and initialize DPUTools
+    args = parser.parse_args()
```
It might be nice to have this main() function call out to another function to set up all the above arguments. The way it is now makes the code a bit harder to read / harder to track where subsequent changes need to be added.

So main would essentially do:
- set up arguments
-  set up logging
-  create dpuTools
- dispatch dpuTools

Any then the more granular argument organization could go in the "set up arguments" function.

SalDaniele commented on utils/common_bf.py:
```diff
@@ -0,0 +1,188 @@
+
+
+@dataclasses.dataclass(frozen=True)
+class Result:
```
can't we just take this from common.py?

SalDaniele reviewed (commented):
Overall, I am happy to merge if @bn222 approves.

From a user-perspective, I feel everything is working and intuitive (i.e. help for each subcommand provides good output).

From a maintainers persepctive, this could use a little more work to make it easy to extend this in the future for whatever DPUs we add support for moving forward.

bn222 commented:
I'm also happy with the changes. We will need to specify which dpu type we're using since we can't detect it.

SamD2021 committed:
requirements.txt: Add all new requirements to allow for bare metal use as well

SamD2021 committed:
dpu-tools: Update ipu reset command with up to date command

- Requires the user to provide the imc address, like with the firmware
command, if the dpu-tools are in ipu mode.
- The new reset command uses that imc address to call reboot on the imc.

SamD2021 committed:
dpu-tools: Let the user override the dpu_type

- It either takes the dpu_type given by the user or auto detects it
otherwise.

SamD2021 committed:
dpu-tools: refactor DPU Tools into an Abstract Base Class

SalDaniele commented:
Nice, I really like the changes, looks good to me. @bn222 can we merge this?
